{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b592450",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing libraries\n",
    "import pandas as pd\n",
    "import os\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9dded322",
   "metadata": {},
   "outputs": [],
   "source": [
    "code = {'TTT' : 'F', 'TTC' : 'F', 'TTA' : 'L', 'TTG' : 'L', \n",
    "        'TCT' : 'S', 'TCC' : 'S', 'TCA' : 'S', 'TCG' : 'S', \n",
    "        'TAT' : 'Y', 'TAC' : 'Y', 'TAA' : '*', 'TAG' : '*', \n",
    "        'TGT' : 'C', 'TGC' : 'C', 'TGA' : '*', 'TGG' : 'W', \n",
    "        'CTT' : 'L', 'CTC' : 'L', 'CTA' : 'L', 'CTG' : 'L', \n",
    "        'CCT' : 'P', 'CCC' : 'P', 'CCA' : 'P', 'CCG' : 'P', \n",
    "        'CAT' : 'H', 'CAC' : 'H', 'CAA' : 'Q', 'CAG' : 'Q', \n",
    "        'CGT' : 'R', 'CGC' : 'R', 'CGA' : 'R', 'CGG' : 'R', \n",
    "        'ATT' : 'I', 'ATC' : 'I', 'ATA' : 'I', 'ATG' : 'M', \n",
    "        'ACT' : 'T', 'ACC' : 'T', 'ACA' : 'T', 'ACG' : 'T', \n",
    "        'AAT' : 'N', 'AAC' : 'N', 'AAA' : 'K', 'AAG' : 'K', \n",
    "        'AGT' : 'S', 'AGC' : 'S', 'AGA' : 'R', 'AGG' : 'R', \n",
    "        'GTT' : 'V', 'GTC' : 'V', 'GTA' : 'V', 'GTG' : 'V', \n",
    "        'GCT' : 'A', 'GCC' : 'A', 'GCA' : 'A', 'GCG' : 'A', \n",
    "        'GAT' : 'D', 'GAC' : 'D', 'GAA' : 'E', 'GAG' : 'E', \n",
    "        'GGT' : 'G', 'GGC' : 'G', 'GGA' : 'G', 'GGG' : 'G'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4022ae35",
   "metadata": {},
   "outputs": [],
   "source": [
    "def header_remover(file):\n",
    "    \"\"\"\n",
    "    Removes the header from the vep output in tab format, creates a copy of the file without the header\n",
    "    :param file: path to vep output in tab format\n",
    "    :returns: the name of the copy of the file without the header\n",
    "    \"\"\"\n",
    "    bashCommand = 'cat ' + file + ' | grep -v ^## > ' + file +'.cl'\n",
    "    os.system(bashCommand)\n",
    "    return file + '.cl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e1eb2b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def input_reader(file):\n",
    "    \"\"\"\n",
    "    Reads a file in tab formal without a header\n",
    "    :param file: path to the input file\n",
    "    :returns: the data from the file as pandas.DataFrame\n",
    "    \"\"\"\n",
    "    out = pd.read_csv(file, sep='\\t', engine='python')\n",
    "    bashCommand = 'rm ' + file\n",
    "    os.system(bashCommand) #removing the file without the header\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c8e933fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def close_finder(table):\n",
    "    \"\"\"\n",
    "    Finds variations in the same codon\n",
    "    :param table: a pandas.DataFrame containing vep annotation data\n",
    "    :returns: a pandas.DataFrame with variations in the same codon\n",
    "    \"\"\"\n",
    "    out = pd.DataFrame(columns=table.columns)\n",
    "    table = table.sort_values('Gene')\n",
    "    genes = list(table['Gene'].unique())\n",
    "    for i in genes:\n",
    "        positions = []\n",
    "        gene_table = table.loc[table['Gene'] == i]\n",
    "        gene_table = gene_table.sort_values('Protein_position')\n",
    "        if gene_table.shape[0] >= 2:\n",
    "            for j in range(gene_table.shape[0]-1):\n",
    "                if gene_table.iloc[j]['Protein_position'] == gene_table.iloc[j+1]['Protein_position']:\n",
    "                    if gene_table.iloc[j]['STRAND'] == gene_table.iloc[j+1]['STRAND']:\n",
    "                        positions.append(gene_table.iloc[j]['Protein_position'])\n",
    "                        positions.append(gene_table.iloc[j+1]['Protein_position'])\n",
    "        gene_table = gene_table[gene_table['Protein_position'].isin(positions)]\n",
    "        out = pd.concat([out, gene_table])\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "27bbaae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MNF_remover(table):\n",
    "    \"\"\"\n",
    "    Removes repeats and misannotated MNVs from the vep annotation data\n",
    "    :param table: a pandas.DataFrame with vep annotation\n",
    "    :returns: a pandas.DataFrame with the vep annotation data without repeats and misannotated MNVs\n",
    "    \"\"\"\n",
    "    return pd.concat([table,close_finder(table)]).drop_duplicates(keep=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fccf5638",
   "metadata": {},
   "outputs": [],
   "source": [
    "def consequences(table):\n",
    "    \"\"\"\n",
    "    Determines the resulting codon, amino acid, consequences and inpact of the found MNVs\n",
    "    :param table: a pandas.DataFrame with MNVs\n",
    "    :returns: annotated pandas.DataFrame with MNVs\n",
    "    \"\"\"\n",
    "    out = pd.DataFrame(columns=table.columns)\n",
    "    table = table.sort_values('Gene')\n",
    "    genes = list(table['Gene'].unique())\n",
    "    \n",
    "    for i in genes:\n",
    "        gene_table = table.loc[table['Gene'] == i]\n",
    "        reference_aa = gene_table.iloc[0]['Amino_acids'][-1]\n",
    "        reference_codon = list(gene_table.iloc[0]['Codons'].split('/')[1].lower())\n",
    "        position = gene_table.iloc[0]['Protein_position']\n",
    "        codon = list(gene_table.iloc[0]['Codons'].split('/')[1].lower())\n",
    "        for j in range(gene_table.shape[0]):\n",
    "            mutant_codon = list(gene_table.iloc[j]['Codons'][0:3])\n",
    "            for k in range(3):\n",
    "                if mutant_codon[k].isupper():\n",
    "                    codon[k] = mutant_codon[k]\n",
    "        score = 0\n",
    "        for c in range(3):\n",
    "            if codon[c].isupper() and codon[c].lower() != reference_codon[c].lower():\n",
    "                score += 1\n",
    "                \n",
    "        if score > 1:\n",
    "            codons = ''.join(codon)+'/'+''.join(reference_codon)\n",
    "            new_table = gene_table.iloc[:1]\n",
    "            new_table['Codons'] = codons\n",
    "            new_aa = code[(''.join(codon)).upper()]\n",
    "            aa = new_aa + '/' + reference_aa\n",
    "            new_table['Amino_acid'] = aa\n",
    "            \n",
    "            # Calculating consequence\n",
    "            if new_aa == reference_aa:\n",
    "                consequence = 'synonymous_variant'\n",
    "            elif new_aa == '*':\n",
    "                consequence = 'stop_gained'\n",
    "            elif reference_aa == '*':\n",
    "                consequence = 'stop_lost'\n",
    "            elif position == 1:\n",
    "                consequence = 'start_lost'\n",
    "            else:\n",
    "                consequence = 'missense_variant'\n",
    "            new_table['Consequence'] = consequence\n",
    "            \n",
    "            # Calculating impact\n",
    "            if consequence == 'synonymous_variant':\n",
    "                impact = 'LOW'\n",
    "            elif consequence == 'missense_variant':\n",
    "                impact = 'MODERATE'\n",
    "            else:\n",
    "                impact = 'HIGH'\n",
    "            new_table['IMPACT'] = impact\n",
    "            \n",
    "            out = pd.concat([out, new_table])\n",
    "        \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "aaef0362",
   "metadata": {},
   "outputs": [],
   "source": [
    "def output_writer(table, path):\n",
    "    \"\"\"\n",
    "    Writes annotation in tab format\n",
    "    :param table: a pandas.DataFrame with annotation\n",
    "    :param path: the directory for the output \n",
    "    \"\"\"\n",
    "    table.to_csv(path, sep ='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "df6426a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MNV_pipeline(file, outpath):\n",
    "    \"\"\"\n",
    "    Combines reading of the data, searching for polymorphisms in the same codon, annotation of MNVs and saving the results in tab format.\n",
    "    There is an option to save SNPs annotation without repeats and misannotated MNVs\n",
    "    :param table: a pandas.DataFrame with annotation\n",
    "    :param path: the directory for the output\n",
    "    \"\"\"\n",
    "    cleaned_file = header_remover(file)\n",
    "    print('Reading data...')\n",
    "    data = input_reader(cleaned_file)\n",
    "    print('Searching for polymorphisms in the same codon...')\n",
    "    close_data = close_finder(data)\n",
    "    snp_out = input('Would like to save file with SNP annotation without misannotated MNVs? y/n ')\n",
    "    if snp_out == 'y':\n",
    "        snp_data = MNF_remover(data)\n",
    "        path = outpath + '/snps.tab'\n",
    "        output_writer(snp_data, path)\n",
    "    print('Annotating MNVs...')\n",
    "    mnv_data = consequences(close_data)\n",
    "    path = outpath + '/mnvs.tab'\n",
    "    output_writer(mnv_data, path)\n",
    "    print('Succesfully finished')\n",
    "    print('Found', mnv_data.shape[0], 'MNVs:')\n",
    "    consequence = mnv_data['Consequence'].value_counts().to_dict()\n",
    "    for key in consequence:\n",
    "        print(key, consequence[key])\n",
    "    print('The output file(s) can be found in ', outpath)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
